### ISSUE TITLE
Concurrent map read & write

### ISSUE THREAD

**Issue by Nv7-GitHub** [pocketbase/pocketbase#1628]
A lot of requests were sent to our server at once, and it crashed. Error log:
```
fatal error: concurrent map iteration and map write

goroutine 15654 [running]:
runtime.throw({0x12db28b?, 0x4000cc1800?})
	/snap/go/10038/src/runtime/panic.go:992 +0x50 fp=0x4000aa66d0 sp=0x4000aa66a0 pc=0x43bc80
runtime.mapiternext(0x4000aa68c8?)
	/snap/go/10038/src/runtime/map.go:871 +0x53c fp=0x4000aa6740 sp=0x4000aa66d0 pc=0x41732c
github.com/pocketbase/pocketbase/apis.(*realtimeApi).broadcastRecord(0x400046c4c0, {0x12aaa3c, 0x6}, 0x4000b0a680)
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/realtime.go:372 +0x5e8 fp=0x4000aa6ac0 sp=0x4000aa6740 pc=0xedb048
github.com/pocketbase/pocketbase/apis.(*realtimeApi).bindEvents.func4(0x4000af5f30?)
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/realtime.go:269 +0x4c fp=0x4000aa6af0 sp=0x4000aa6ac0 pc=0xeda1dc
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger(0x400028a9f0?, 0x40009e8c00, {0x0?, 0x0, 0x4000ac8160})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/tools/hook/hook.go:65 +0x21c fp=0x4000aa6b80 sp=0x4000aa6af0 pc=0xea754c
github.com/pocketbase/pocketbase/core.(*BaseApp).createDaoWithHooks.func4(0x400018bc00, {0x1cf2520?, 0x4000b0a680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/core/base.go:913 +0xc0 fp=0x4000aa6bf0 sp=0x4000aa6b80 pc=0xea5e60
github.com/pocketbase/pocketbase/daos.(*Dao).update(0x400018bc00, {0x1cf2520, 0x4000b0a680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/daos/base.go:295 +0x234 fp=0x4000aa6c80 sp=0x4000aa6bf0 pc=0xe776f4
github.com/pocketbase/pocketbase/daos.(*Dao).Save.func2(0x40008dc390?)
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/daos/base.go:255 +0x28 fp=0x4000aa6cb0 sp=0x4000aa6c80 pc=0xe77448
github.com/pocketbase/pocketbase/daos.(*Dao).failRetry(0x400018bc00, 0x4000aa6d50, 0x5)
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/daos/base.go:365 +0x184 fp=0x4000aa6d20 sp=0x4000aa6cb0 pc=0xe77c24
github.com/pocketbase/pocketbase/daos.(*Dao).Save(0x1092620?, {0x1cf2520, 0x4000b0a680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/daos/base.go:254 +0xac fp=0x4000aa6d70 sp=0x4000aa6d20 pc=0xe773ec
github.com/pocketbase/pocketbase/daos.(*Dao).SaveRecord(0x4000602540?, 0x4000b0a680?)
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/daos/record.go:344 +0x110 fp=0x4000aa6e10 sp=0x4000aa6d70 pc=0xe7d1f0
github.com/pocketbase/pocketbase/forms.(*RecordUpsert).Submit.func1()
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/forms/record_upsert.go:737 +0xa8 fp=0x4000aa6f70 sp=0x4000aa6e10 pc=0xeca318
github.com/pocketbase/pocketbase/apis.(*recordApi).update.func2.1.1(0x4000620f18)
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/record_crud.go:314 +0x34 fp=0x4000aa7000 sp=0x4000aa6f70 pc=0xee27b4
github.com/pocketbase/pocketbase/tools/hook.(*Hook[...]).Trigger(0x40001a07e0?, 0x7101, {0x4000aa70e8?, 0x1, 0xebd078})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/tools/hook/hook.go:65 +0x21c fp=0x4000aa7090 sp=0x4000aa7000 pc=0xea754c
github.com/pocketbase/pocketbase/apis.(*recordApi).update.func2.1()
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/record_crud.go:313 +0xc4 fp=0x4000aa7100 sp=0x4000aa7090 pc=0xee2764
github.com/pocketbase/pocketbase/forms.runInterceptors(0x4000602540?, {0x4000aa7200, 0x1?, 0xffffffffffffffff?})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/forms/base.go:27 +0x54 fp=0x4000aa7120 sp=0x4000aa7100 pc=0xebd094
github.com/pocketbase/pocketbase/forms.(*RecordUpsert).Submit(0x4000602540, {0x4000aa7200, 0x1, 0x1})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/forms/record_upsert.go:725 +0x84 fp=0x4000aa7150 sp=0x4000aa7120 pc=0xeca224
github.com/pocketbase/pocketbase/apis.(*recordApi).update(0x4000531a80, {0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/record_crud.go:311 +0x4d0 fp=0x4000aa72e0 sp=0x4000aa7150 pc=0xee24c0
github.com/pocketbase/pocketbase/apis.(*recordApi).update-fm({0x1cf5d38?, 0x40009e0680?})
	<autogenerated>:1 +0x40 fp=0x4000aa7310 sp=0x4000aa72e0 pc=0xee7470
github.com/pocketbase/pocketbase/apis.LoadCollectionContext.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/middlewares.go:270 +0x140 fp=0x4000aa73b0 sp=0x4000aa7310 pc=0xed71c0
github.com/pocketbase/pocketbase/apis.ActivityLogger.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/middlewares.go:283 +0x48 fp=0x4000aa74b0 sp=0x4000aa73b0 pc=0xed7438
github.com/labstack/echo/v5/middleware.CORSConfig.ToMiddleware.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20220201181537-ed2888cfa198/middleware/cors.go:208 +0x74c fp=0x4000aa76a0 sp=0x4000aa74b0 pc=0x74a88c
github.com/pocketbase/pocketbase/apis.LoadAuthContext.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/pocketbase/pocketbase@v0.11.2/apis/middlewares.go:245 +0x238 fp=0x4000aa7720 sp=0x4000aa76a0 pc=0xed6e28
github.com/labstack/echo/v5/middleware.SecureConfig.ToMiddleware.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20220201181537-ed2888cfa198/middleware/secure.go:142 +0x7d4 fp=0x4000aa78a0 sp=0x4000aa7720 pc=0x74c8e4
github.com/labstack/echo/v5/middleware.RecoverConfig.ToMiddleware.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20220201181537-ed2888cfa198/middleware/recover.go:78 +0xf8 fp=0x4000aa7920 sp=0x4000aa78a0 pc=0x74b618
github.com/labstack/echo/v5.(*Echo).ServeHTTP.func1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20220201181537-ed2888cfa198/echo.go:620 +0x118 fp=0x4000aa7980 sp=0x4000aa7920 pc=0x6ec3e8
github.com/labstack/echo/v5/middleware.RemoveTrailingSlashConfig.ToMiddleware.func1.1({0x1cf5d38, 0x40009e0680})
	/home/pi/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20220201181537-ed2888cfa198/middleware/slash.go:136 +0x1d8 fp=0x4000aa7a10 sp=0x4000aa7980 pc=0x74cd48
github.com/labstack/echo/v5.(*Echo).ServeHTTP(0x4000193500, {0x1ced808, 0x40001402a0}, 0x40009da400)
	/home/pi/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20220201181537-ed2888cfa198/echo.go:626 +0x218 fp=0x4000aa7a80 sp=0x4000aa7a10 pc=0x6ec218
net/http.serverHandler.ServeHTTP({0x1ceb618?}, {0x1ced808, 0x40001402a0}, 0x40009da400)
	/snap/go/10038/src/net/http/server.go:2916 +0x408 fp=0x4000aa7b40 sp=0x4000aa7a80 pc=0x697478
net/http.(*conn).serve(0x4000bfa1e0, {0x1cedfa8, 0x4000330cc0})
	/snap/go/10038/src/net/http/server.go:1966 +0x550 fp=0x4000aa7fa0 sp=0x4000aa7b40 pc=0x692fd0
net/http.(*Server).Serve.func3()
	/snap/go/10038/src/net/http/server.go:3071 +0x34 fp=0x4000aa7fd0 sp=0x4000aa7fa0 pc=0x697cc4
runtime.goexit()
	/snap/go/10038/src/runtime/asm_arm64.s:1270 +0x4 fp=0x4000aa7fd0 sp=0x4000aa7fd0 pc=0x46cac4
created by net/http.(*Server).Serve
	/snap/go/10038/src/net/http/server.go:3071 +0x448
```
(there were logs of every other goroutine but I left them out since most were just waiting)

**Comment by ganigeorgiev** [comment]
Hm, this seems similar to https://github.com/pocketbase/pocketbase/issues/1566.

I think it is related to the fact that we don't have mutex for `models.Record` and some operations may not be concurrent safe. I'll have a more detailed look at it sometime later today and will try to provide a fix.

**Comment by ganigeorgiev** [comment]
The realtime panic should be fixed with the latest [v0.11.3 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.11.3).

_`models.Record` is still not concurrent safe because I have some performance concerns and I'll have to test it more thoroughly and for now is left v0.12/v0.13._

**Comment by ganigeorgiev** [comment]
_Just for info if anyone is following here, the `models.Record` data and expand operations should be now concurrent safe in the latest [v0.12.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.12.0)._
